### DISCUSSION TITLE
I don't understand what is meant by needing a virtual environment

### DISCUSSION THREAD

**Discussion by rfalanga** [skills/secure-code-game#99] [Q&A] [answered]
I'm still working my way through this training course because I want to learn it. I've got limited experience with Python, and no experience using it in Codespaces. (I'm still new to Codespaces.) When I tried running the code.py file, I got a message saying that I don't have a virtual environment set up. This raises multiple questions for me:

- I added the Python extension when I started this lab. I thought that it would have created whatever Python environment I needed, but it doesn't look like it did. What environment should I choose?
- How much do these Python virtual environments cost?
- Do they time out, like Codespaces do, after 30 minutes of inactivity?

**Comment by jkcso** [comment, marked as answer]
Click 'no' in that question about virtual environment and proceed with the course. The `tests.py` should run as normal.
There is no cost associated with a virtual environment inside Codespaces and you have 60 hours of free usage every month.
Even if a codespace times out, you can resume it. To avoid losing your progress you can `push` your work to GitHub, as normal

**Reply by jkcso** [reply]
I have now updated the repo (8e9fb568055e63b9c2f9b8282e2ec53ea1d4e256) with this info. Thanks for letting us know!
